cirq/protocols/unitary.py

```diff
@@ -98,6 +99,9 @@ def unitary(val: Any,
     """
     from cirq import Gate, Operation  # HACK: Avoids circular dependencies.
 
+    if isinstance(val, np.ndarray) and linalg.is_unitary(val):
```

Comment:
This is a pretty significant change to the semantics of cirq.unitary, not mentioned in the PR description.

I am very unsure about the "is unitary" check, because it is cubically more expensive than just reading the array. Also, it is technically stricter than what is allowed to come out of other things. Like, calling `cirq.unitary` on a malformed gate would cause no error, but `cirq.unitary(cirq.unitary(bad_gate))` would...?

I do like the idea of "How do I make a unitary thing?" being answered by "Just give the matrix as a numpy array and it will work".

Reply:
Fair enough, I should have mentioned it in the PR description. I ran into the need to introduce this when I was modifying the mixture simulation to use apply_unitary methods for the gates/operations that were returned. My understanding was that a user might still end up defining a channel for themselves that returned `ndarray`s in the mixture. But calling `cirq.unitary` on a `ndarray` throws an error right now. I therefore added a check to see if the val I received IS a unitary and returned it back to the user in that case. The `is_unitary` check only gets executed if `val` is an `ndarray` because the `and` statement will short-circuit. If the user provides a malformed (non-unitary) matrix then `is_unitary` check returns `False` and the user would get the same error a user currently gets for passing in an `ndarray`. Admittedly that error message isn't appropriate and should instead be something along the lines of "The array you gave me isn't a unitary".